README


                 _ __  _   _  ___ _ __ ___
                | '_ \| | | |/ __| '__/ __|
                | |_) | |_| | (__| | | (__
                | .__/ \__, |\___|_|  \___|
                |_|    |___/

                http://www.tty1.net/pycrc/


pycrc is a free, easy to use Cyclic Redundancy Check (CRC) calculator and
C source code generator.



System Requirements
===================

pycrc requires Python 2.6 or later. Python 3.x is supported.



Installation
============

pycrc requires Python 2.6 or later. Python 3.x is supported.

This program doesn't need any particular installation. The script can be
called from any directory.
Simply call the python interpreter with the script as parameter:

python pycrc.py [options]

On UNIX-like systems, you might want to make the script executable:

chmod +x pycrc.py

Then the script can be called like an application.
./pycrc.py [options]



Getting help
============

If you are new to pycrc and want to generate C code, start with
[the tutorial](http://www.tty1.net/pycrc/tutorial_en.html).

The [pycrc manual page](http://www.tty1.net/pycrc/pycrc.html) explains the command
line options in some detail and also gives some more examples how to use pycrc.

If you have questions about using pycrc which is not answered in a satisfactory
way by the documentation, please send a mail to the pycrc user mailing list
<pycrc-users@lists.sourceforge.net>. Subscribe to the mailing list
[pycrc users list](https://lists.sourceforge.net/lists/listinfo/pycrc-users).
Due to excessive spamming a subscription is required to post on the list.

If you have found a bug in pycrc, please take the time and file it to the
[bug tracker](http://sourceforge.net/p/pycrc/bugs/) or record a
[feature request](http://sourceforge.net/p/pycrc/feature-requests/).
Thanks for your collaboration.

Also see the [frequently asked questions](http://www.tty1.net/pycrc/faq.html).



Feedback
========

If you like pycrc, let me know and drop me a note. If you don't like pycrc let
me know what you don't like and why. In both cases, I would really appreciate
some [feed back](http://sourceforge.net/projects/pycrc/reviews/).
If you want some idea how to say thanks for this software, please have a look
[here](http://www.tty1.net/say-thanks_en.html).



Copyright of the generated source code
======================================

Prior to v0.6, pycrc was released under the GPL and an additional addition to
the licence was required to permit to use the generated source code in products
with a OSI unapproved licence. As of version 0.6, pycrc is released under the
terms of the MIT licence and such an additional clause to the licence is no
more required.


ChangeLog
2016-01-16  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * pycrc.py:
    Fixed binary files handling with Python 2.7.
    Fixes #11. Thanks to James Bowman.

    * doc/pycrc.xml:
    Corrected a typo.

2016-01-06  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_opt.py:
    Version 0.9

2015-10-31  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_symtable.py:
    Suppressed the crc_reflect function where not needed. Addresses part of #8.
    Thanks to Craig McQueen.

2015-10-26  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * pycrc.py:
    Allow strings with values greater than 0x80 in --check-hexstring.

    * crc_algorithms.py:
    When the CRC width is less than 8 then the bit-by-bit algorithm needs to
    apply the CRC mask to the final value.
    Fixes #10. Thanks to Steve Geo.

2015-10-25  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_models.py:
    * doc/pycrc.xml:
    Fixed the initial value of the 16-bit CCITT algorithm. Renamed the model
    from ccitt to crc-16-ccitt.
    Fixes #7. Thanks to Craig McQueen.

2015-10-11  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * doc/pycrc.xml:
    Documented the experimental --slice-by option.

2015-10-06  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_algorithms.py:
    Simplified the implementation where Width is less than 8 bits.

2015-10-06  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * AUTHORS:
    Added Stephan Brumme as his implementation of the slice-by algorithm is the
    basis for pycrc's implementation.

    * crc_algorithms.py:
    * crc_opt.py:
    * crc_symtable.py:
    Added a new option --slice-by. This option is still experimental and
    limited in its use.

    * test/test.py:
    Added tests for --slice-by.

2015-09-12  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_algorithms.py:
    * crc_lexer.py:
    * crc_models.py:
    * crc_opt.py:
    * crc_parser.py:
    * crc_symtable.py:
    * pycrc.py:
    Run the code through pylint.
    API change: changed the names of the member variables from CamelCase to the
    format suggested in PEP 0008 (lowercase letters and words separated by
    underscore).

2015-08-31  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_opt.py:
    Version 0.8.3

2015-08-29  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * README.md:
    * doc/pycrc.xml:
    pycrc has a new homepage: https://pycrc.org

2015-08-12  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_symtable.py:
    * test/performance.sh:
    Applied some minor optimisations to the generated table-driven code.

    * AUTHORS:
    Belatedly added an authors file. If I have forgotten to mention someone
    please don't hesitate to send a mail.

2015-08-11  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * doc/pycrc.xml:
    * doc/Makefile:
    Upgraded documentation to DocBook 5.

    * README.md:
    Removed sourceforge mailing list from README.md in an effort to move pycrc
    away from sourceforge.

2015-07-12  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_opt.py:
    * crc_symtable.py:
    * doc/pycrc.xml:
    * qm.py:
    * test/check_files.sh:
    * test/performance.sh:
    * test/test.py:
    Removed the experimental --bitwise-expression option to facilitate
    restructuring of the code.

2015-04-04  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * pycrc.py:
    * crc_algorithms.py:
    * test/test.py:
    The preferred format for the input data for the Python API is now byte
    array. But if a string is supplied it is decoded as UTF-8 string.
    Alternative formats are not supported and must be passed to the functions
    as byte arrays.

2015-01-21  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * README.md:
    The issue tracker on GitHub https://github.com/tpircher/pycrc/issues
    is now advertised as the preferred issue tracker.

2015-01-20  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_symtable.py:
    Changed the signature of the crc_update() function: the data argument is
    now a pointer to void to improve compatibility with C++.
    Thanks to Kamil Szczygieł.
    This closes GitHub issue #4.

2014-12-04  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_opt.py:
    Version 0.8.2

2014-12-04  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_symtable.py:
    Smaller code cleanups.

    * doc/pyxrx.xml:
    Stated more clearly that the bitwise expression algorithm is experimental.

2014-07-18  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * doc/pyxrx.xml:
    Fixed a typo in the documentation.
    The description previously stated:
        The reflected value of 0xa3 (10100010b) is 0x45 (01000101b)
    but this should be:
        The reflected value of 0xa2 (10100010b) is 0x45 (01000101b)
    Thanks to Andreas Nebenfuehr for reporting the mistake.

2013-12-23  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_symtable.py:
    The table-driven code for polynomials of width < 8 using a table index
    width < 8 was producing a wrong checksum.
    Thanks to Radosław Gancarz.

2013-12-23  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_symtable.py:
    Small cleanups.

    * test/test.py:
    Small cleanups.
    Added a tests for special cases. For now, added crc-5 with non-inverted
    input. This test is currently failing.

2013-12-20  Thomas Pircher   <tehpeh-pycrc@tty1.net>

    * crc_symtable.py:
    * test/test.py:
    Updated the generated code to cope with big Widths (>32 bits) on 32 bit
    processors.
    Since C89 does not give a way to specify the minimum length of a data type,

crc_opt.py
#  pycrc -- parameterisable CRC calculation utility and C source code generator
#
#  Copyright (c) 2006-2016  Thomas Pircher  <tehpeh-pycrc@tty1.net>
#
#  Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
#  IN THE SOFTWARE.


"""
Option parsing library for pycrc.
use as follows:

   from crc_opt import Options

   opt = Options()
   opt.parse(sys.argv[1:])
"""

from optparse import OptionParser, Option, OptionValueError
from copy import copy
import sys
from crc_models import CrcModels


# Class Options
###############################################################################
class Options(object):
    """
    The options parsing and validating class.
    """
    # pylint: disable=too-many-instance-attributes, too-few-public-methods

    # Program details
    program_name = "pycrc"
    version = "0.9"
    version_str = "{0:s} v{1:s}".format(program_name, version)
    web_address = "https://pycrc.org"

    # Bitmap of the algorithms
    algo_none = 0x00
    algo_bit_by_bit = 0x01
    algo_bit_by_bit_fast = 0x02
    algo_table_driven = 0x08

    action_check_str = 0x01
    action_check_hex_str = 0x02
    action_check_file = 0x03
    action_generate_h = 0x04
    action_generate_c = 0x05
    action_generate_c_main = 0x06
    action_generate_table = 0x07


    # Class constructor
    ###############################################################################
    def __init__(self):
        self.width = None
        self.poly = None
        self.reflect_in = None
        self.xor_in = None
        self.reflect_out = None
        self.xor_out = None
        self.tbl_idx_width = 8
        self.tbl_width = 1 << self.tbl_idx_width
        self.slice_by = 1
        self.verbose = False
        self.check_string = "123456789"
        self.msb_mask = None
        self.mask = None

        self.algorithm = self.algo_none
        self.symbol_prefix = "crc_"
        self.crc_type = None
        self.include_files = []
        self.output_file = None
        self.action = self.action_check_str
        self.check_file = None
        self.c_std = None
        self.undefined_crc_parameters = False


    # function parse
    ###############################################################################
    def parse(self, argv=None):
        """
        Parses and validates the options given as arguments
        """
        # pylint: disable=too-many-branches, too-many-statements

        usage = """python %prog [OPTIONS]

To calculate the checksum of a string or hexadecimal data:
    python %prog [model] --check-string "123456789"
    python %prog [model] --check-hexstring "313233343536373839"

To calculate the checksum of a file:
    python %prog [model] --check-file filename

To generate the C source code and write it to filename:
    python %prog [model] --generate c -o filename

The model can be defined either with the --model switch or by specifying each
of the following parameters:
    --width --poly --reflect-in --xor-in --reflect-out --xor-out"""

        models = CrcModels()
        model_list = ", ".join(models.get_list())
        parser = OptionParser(option_class=MyOption, usage=usage, version=self.version_str)
        parser.add_option(
            "-v", "--verbose",
            action="store_true", dest="verbose", default=False,
            help="be more verbose; print the value of the parameters "
            "and the chosen model to stdout")
        parser.add_option(
            "--check-string",
            action="store", type="string", dest="check_string",
            help="calculate the checksum of a string (default: '123456789')",
            metavar="STRING")
        parser.add_option(
            "--check-hexstring",
            action="store", type="string", dest="check_hexstring",
            help="calculate the checksum of a hexadecimal number string",
            metavar="STRING")
        parser.add_option(
            "--check-file",
            action="store", type="string", dest="check_file",
            help="calculate the checksum of a file",
            metavar="FILE")
        parser.add_option(
            "--generate",
            action="store", type="string", dest="generate", default=None,
            help="generate C source code; choose the type from {h, c, c-main, table}",
            metavar="CODE")
        parser.add_option(
            "--std",
            action="store", type="string", dest="c_std", default="C99",
            help="choose the C dialect of the generated code from {C89, ANSI, C99}",
            metavar="STD")
        parser.add_option(
            "--algorithm",
            action="store", type="string", dest="algorithm", default="all",
            help="choose an algorithm from "
            "{bit-by-bit, bbb, bit-by-bit-fast, bbf, table-driven, tbl, all}",
            metavar="ALGO")
        parser.add_option(
            "--model",
            action="callback", callback=_model_cb, type="string", dest="model", default=None,
            help="choose a parameter set from {{{0:s}}}".format(model_list),
            metavar="MODEL")
        parser.add_option(
            "--width",
            action="store", type="hex", dest="width",
            help="use NUM bits in the polynomial",
            metavar="NUM")
        parser.add_option(
            "--poly",
            action="store", type="hex", dest="poly",
            help="use HEX as polynomial",
            metavar="HEX")
        parser.add_option(
            "--reflect-in",
            action="store", type="bool", dest="reflect_in",
            help="reflect the octets in the input message",
            metavar="BOOL")
        parser.add_option(
            "--xor-in",
            action="store", type="hex", dest="xor_in",
            help="use HEX as initial value",
            metavar="HEX")
        parser.add_option(
            "--reflect-out",
            action="store", type="bool", dest="reflect_out",
            help="reflect the resulting checksum before applying the --xor-out value",
            metavar="BOOL")
        parser.add_option(
            "--xor-out",
            action="store", type="hex", dest="xor_out",
            help="xor the final CRC value with HEX",
            metavar="HEX")
        parser.add_option(
            "--slice-by",
            action="store", type="int", dest="slice_by",
            help="read NUM bytes at a time from the input. NUM must be one of the values {4, 8, 16}",
            metavar="NUM")
        parser.add_option(
            "--table-idx-width",
            action="store", type="int", dest="table_idx_width",
            help="use NUM bits to index the CRC table; NUM must be one of the values {1, 2, 4, 8}",
            metavar="NUM")
        parser.add_option(
            "--symbol-prefix",
            action="store", type="string", dest="symbol_prefix",
            help="when generating source code, use STRING as prefix to the exported C symbols",
            metavar="STRING")
        parser.add_option(
            "--crc-type",
            action="store", type="string", dest="crc_type",
            help="when generating source code, use STRING as crc_t type",
            metavar="STRING")
        parser.add_option(
            "--include-file",
            action="append", type="string", dest="include_files",
            help="when generating source code, include also FILE as header file; "
            "can be specified multiple times",
            metavar="FILE")
        parser.add_option(
            "-o", "--output",
            action="store", type="string", dest="output_file",
            help="write the generated code to file instead to stdout",
            metavar="FILE")

        (options, args) = parser.parse_args(argv)

        if options.c_std != None:
            std = options.c_std.upper()
            if std == "ANSI" or std == "C89":
                self.c_std = "C89"
            elif std == "C99":
                self.c_std = std
            else:
                self.__error("unknown C standard {0:s}".format(options.c_std))

        undefined_params = []
        if options.width != None:
            self.width = options.width
        else:
            undefined_params.append("--width")
        if options.poly != None:
            self.poly = options.poly
        else:
            undefined_params.append("--poly")
        if options.reflect_in != None:
            self.reflect_in = options.reflect_in
        else:
            undefined_params.append("--reflect-in")
        if options.xor_in != None:
            self.xor_in = options.xor_in
        else:
            undefined_params.append("--xor-in")
        if options.reflect_out != None:
            self.reflect_out = options.reflect_out
        else:
            undefined_params.append("--reflect-out")
        if options.xor_out != None:
            self.xor_out = options.xor_out
        else:
            undefined_params.append("--xor-out")

        if options.table_idx_width != None:
            if options.table_idx_width in set((1, 2, 4, 8)):
                self.tbl_idx_width = options.table_idx_width
                self.tbl_width = 1 << options.table_idx_width
            else:
                self.__error("unsupported table-idx-width {0:d}".format(options.table_idx_width))

        if self.width != None:
            if self.width <= 0:
                self.__error("Width must be strictly positive")
            self.msb_mask = 0x1 << (self.width - 1)
            self.mask = ((self.msb_mask - 1) << 1) | 1
            if self.poly != None:
                self.poly = self.poly & self.mask
            if self.xor_in != None:
                self.xor_in = self.xor_in & self.mask
            if self.xor_out != None:
                self.xor_out = self.xor_out & self.mask
        else:
            self.msb_mask = None
            self.mask = None

        if self.width == None or \
                self.poly == None or \
                self.reflect_in == None or \
                self.xor_in == None or \
                self.reflect_out == None or \
                self.xor_out == None:
            self.undefined_crc_parameters = True
        else:
            self.undefined_crc_parameters = False

        if options.slice_by != None:
            if options.slice_by in set((4, 8, 16)):
                self.slice_by = options.slice_by
            else:
                self.__error("unsupported slice-by {0:d}".format(options.slice_by))
            if self.undefined_crc_parameters:
                self.__error("slice-by is only implemented for fully defined models")
            if self.tbl_idx_width != 8:
                self.__error("slice-by is only implemented for table-idx-width=8")
            # FIXME tp: Fix corner cases and disable the following tests
            if self.width < 8:
                self.__warning("disabling slice-by for width {0}".format(self.width))
                self.slice_by = 1
            if self.width < 16:
                self.__warning("disabling slice-by for width {0}".format(self.width))
                self.slice_by = 1
            if self.width > 32:
                self.__warning("disabling slice-by for width {0}".format(self.width))
                self.slice_by = 1
            if not self.reflect_in:
                self.__warning("disabling slice-by for non-reflected algorithm")
                self.slice_by = 1
# FIXME tp: reintroduce this?
#            if self.width % 8 != 0:
#                self.__error("slice-by is only implemented for width multiples of 8")
#            if options.slice_by < self.width / 8:
#                self.__error("slice-by must be greater or equal width / 8")
            if self.c_std == "C89":
                self.__error("--slice-by not supported for C89")

        if options.algorithm != None:
            alg = options.algorithm.lower()
            if alg in set(["bit-by-bit", "bbb", "all"]):
                self.algorithm |= self.algo_bit_by_bit
            if alg in set(["bit-by-bit-fast", "bbf", "all"]):
                self.algorithm |= self.algo_bit_by_bit_fast
            if alg in set(["table-driven", "tbl", "all"]):
                self.algorithm |= self.algo_table_driven
            if self.algorithm == 0:
                self.__error("unknown algorithm {0:s}".format(options.algorithm))

        if options.symbol_prefix != None:
            self.symbol_prefix = options.symbol_prefix
        if options.include_files != None:
            self.include_files = options.include_files
        if options.crc_type != None:
            self.crc_type = options.crc_type
        if options.output_file != None:
            self.output_file = options.output_file
        op_count = 0
        if options.check_string != None:
            self.action = self.action_check_str
            self.check_string = options.check_string
            op_count += 1
        if options.check_hexstring != None:
            self.action = self.action_check_hex_str
            self.check_string = options.check_hexstring
            op_count += 1
        if options.check_file != None:
            self.action = self.action_check_file
            self.check_file = options.check_file
            op_count += 1
        if options.generate != None:
            arg = options.generate.lower()
            if arg == 'h':
                self.action = self.action_generate_h
            elif arg == 'c':
                self.action = self.action_generate_c
            elif arg == 'c-main':
                self.action = self.action_generate_c_main
            elif arg == 'table':
                self.action = self.action_generate_table
            else:
                self.__error("don't know how to generate {0:s}".format(options.generate))
            op_count += 1

            if self.action == self.action_generate_table:
                if self.algorithm & self.algo_table_driven == 0:
                    self.__error("the --generate table option is incompatible "
                        "with the --algorithm option")
                self.algorithm = self.algo_table_driven
            elif self.algorithm not in set(
                    [self.algo_bit_by_bit, self.algo_bit_by_bit_fast, self.algo_table_driven]):
                self.__error("select an algorithm to be used in the generated file")
        else:
            if self.tbl_idx_width != 8:
                self.__warning("reverting to Table Index Width = 8 "
                    "for internal CRC calculation")
                self.tbl_idx_width = 8
                self.tbl_width = 1 << options.table_idx_width
        if op_count == 0:
            self.action = self.action_check_str
        if op_count > 1:
            self.__error("too many actions scecified")

        if len(args) != 0:
            self.__error("unrecognized argument(s): {0:s}".format(" ".join(args)))

        def_params_acts = (self.action_check_str, self.action_check_hex_str,
                           self.action_check_file, self.action_generate_table)
        if self.undefined_crc_parameters and self.action in set(def_params_acts):
            self.__error("undefined parameters: Add {0:s} or use --model"
                .format(", ".join(undefined_params)))
        self.verbose = options.verbose



    # function __warning
    ###############################################################################
    def __warning(self, message):
        """
        Print a warning message to stderr.
        """
        sys.stderr.write(
            "{0:s}: warning: {1:s}\n".format(sys.argv[0], message))



    # function _error
    ###############################################################################
    def __error(self, message):
        """
        Print a error message to stderr and terminate the program.
        """
        sys.stderr.write(
            "{0:s}: error: {1:s}\n".format(sys.argv[0], message))
        sys.exit(1)


# function _model_cb
##############################################################################
def _model_cb(option, opt_str, value, parser):
    """
    This function sets up the single parameters if the 'model' option has been selected
    by the user.
    """
    model_name = value.lower()
    models = CrcModels()
    model = models.get_params(model_name)
    if model != None:
        setattr(parser.values, 'width', model['width'])
        setattr(parser.values, 'poly', model['poly'])
        setattr(parser.values, 'reflect_in', model['reflect_in'])
        setattr(parser.values, 'xor_in', model['xor_in'])
        setattr(parser.values, 'reflect_out', model['reflect_out'])
        setattr(parser.values, 'xor_out', model['xor_out'])
    else:
        models = CrcModels()
        model_list = ", ".join(models.get_list())
        raise OptionValueError(
            "unsupported model {0:s}. Supported models are: {1:s}."
            .format(value, model_list))


# function _check_hex
###############################################################################
def _check_hex(dummy_option, opt, value):
    """
    Checks if a value is given in a decimal integer of hexadecimal reppresentation.
    Returns the converted value or rises an exception on error.
    """
    try:
        if value.lower().startswith("0x"):
            return int(value, 16)
        else:
            return int(value)
    except ValueError:
        raise OptionValueError(
            "option {0:s}: invalid integer or hexadecimal value: {1:s}.".format(opt, value))


# function _check_bool
###############################################################################
def _check_bool(dummy_option, opt, value):
    """
    Checks if a value is given as a boolean value (either 0 or 1 or "true" or "false")
    Returns the converted value or rises an exception on error.
    """
    if value.isdigit():
        return int(value, 10) != 0
    elif value.lower() == "false":
        return False
    elif value.lower() == "true":
        return True
    else:
        raise OptionValueError("option {0:s}: invalid boolean value: {1:s}.".format(opt, value))


# Class MyOption
###############################################################################
class MyOption(Option):
    """
    New option parsing class extends the Option class
    """
    TYPES = Option.TYPES + ("hex", "bool")
    TYPE_CHECKER = copy(Option.TYPE_CHECKER)
    TYPE_CHECKER["hex"] = _check_hex
    TYPE_CHECKER["bool"] = _check_bool

